### ISSUE TITLE
Extended resources failing to register prevent Karpenter to reclaim instances

### ISSUE THREAD

**Issue by cebernardi** [aws/karpenter-core#341]
### Version

__Karpenter Version:__ v0.25.0

__Kubernetes Version:__ v1.23.0

### Expected Behavior

We have a loose Karpenter provisioner that allows for a broad range of machines.

We noticed that sometimes Karpenter spins up `inf1` machines, as they represent a cost effective option at the moment of the evaluation. 

Since we do not provision any AWS Neuron Kubernetes plugin, the extended resource fail to register, but the machine successfully joins the cluster, and is visible to the kubernetes scheduler, which is able to schedule pods on the instance.

We expect that Karpenter will manage the lifecycle of the node normally, reclaiming if empty, terminating it when the TTL expires, and considering the node for the consolidation process.

### Actual Behavior

Since the extended resource is not registered, the node fails the initialisation checks and the the is not marked with the initialisation label `karpenter.sh/initialized=true`:

https://github.com/aws/karpenter-core/blob/8caf21a75aa8914e978e5368c6385d2a2a9d440a/pkg/controllers/machine/lifecycle/initialization.go#L68

Missing this label, the node is not handled for:
- emptiness (https://github.com/aws/karpenter-core/blob/8caf21a75aa8914e978e5368c6385d2a2a9d440a/pkg/controllers/node/emptiness.go#L50)
- deprovisioning: (https://github.com/aws/karpenter-core/blob/8caf21a75aa8914e978e5368c6385d2a2a9d440a/pkg/controllers/deprovisioning/types.go#L96-L101)
- inflight capacity calculation

### Steps to Reproduce the Problem

Our case involved `inf1` machines, but the same will happen with any not handled extended resource.

To reproduce using `inf1` machine family:

- create a loose provisioner with a short TTL

```
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  name: default
spec:
  consolidation:
    enabled: true
  ttlSecondsUntilExpired: 180
```

- deploy a pod selecting for `inf1` family
```
nodeSelector:
  karpenter.k8s.aws/instance-family: inf1
  karpenter.sh/provisioner-name: default
```
(the pod should not specify any `aws.amazon.com/neuron` resource).

The `inf1` instance successfully joins the cluster, Karpenter attaches an event to the node (`Expected resource "aws.amazon.com/neuron" didn't register on the node`), and the pod is scheduled.

however, is not labeled as initialised:

```
kubectl get node -L karpenter.sh/initialized -L karpenter.sh/provisioner-name

NAME                                          STATUS   ROLES    AGE     VERSION                INITIALIZED   PROVISIONER-NAME
ip-x-x-xx.compute.internal        Ready    <none>    x           x                                                     default
```

The node is not reclaimed when the TTL expires, and if the pod is deleted, is not reclaimed for emptiness

### Resource Specs and Logs

Provisioner spec:
```
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  name: general-provisioner
spec:
  consolidation:
    enabled: true
  labels:
    karpenter-provisioner-type: standard
  providerRef:
    name: <template>
  requirements:
  - key: topology.kubernetes.io/zone
    operator: In
    values:
    - us-west-2a
    - us-west-2b
    - us-west-2c
  - key: kubernetes.io/arch
    operator: In
    values:
    - amd64
    - arm64
  - key: karpenter.sh/capacity-type
    operator: In
    values:
    - on-demand
    - spot
  - key: karpenter.k8s.aws/instance-size
    operator: In
    values:
    - 4xlarge
    - 6xlarge
    - 8xlarge
    - 9xlarge
    - 10xlarge
    - 12xlarge
    - 16xlarge
    - 18xlarge
    - 24xlarge
    - 32xlarge
    - 48xlarge
  - key: karpenter.k8s.aws/instance-category
    operator: NotIn
    values:
    - t
    - a
    - x
    - h
  - key: kubernetes.io/os
    operator: In
    values:
    - linux
  ttlSecondsUntilExpired: 2592000
```

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by bwagner5** [comment]
As Todd mentioned here https://github.com/aws/karpenter/issues/3910#issuecomment-1557428018. I'd recommend the same of installing the neuron device plugin or excluding inf from the provisioner. It looks like inf's wouldn't be economical the vast majority of the time. The spot price is generally quite a bit higher than comparable instance types.  Do you run the NVIDIA device plugin for other GPU types? Or do you have the same issue?

**Comment by bwagner5** [comment]
reposting @jonathan-innis response from https://github.com/aws/karpenter/issues/3910#issuecomment-1557996165

We have updated the initialization logic starting in #3408. This change means that we would only wait for the extended resources to register that we have simulated based on pod requests. e.g. if you have a set of pods that don't request neuron and then launch an inf instance, we won't wait for the neuron extended resource to register anymore.

This new logic was released in the release candidate v0.28.0-rc.1 which should get promoted to stable v0.28.0 sometime this week. If you want to try the release candidate in the meantime, you could see if it solves the issue you are running into with inf instances.

**Comment by cebernardi** [comment]
thank you for your reply, we run indeed other gpu plugins (not neuron). 
I understand from what you're writing, that this behaviour is not a bug and is intentional, and part of it is addressed in the candidate release.
Thank you, closing.
